### PR TITLE
[ODS-5058] EdOrg Strategies - Implement code changes to redirect Filters to use the new tuple table only

### PR DESCRIPTION
**Pull request by semalaiappan**

### CODE REVIEW

File: Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/ICriteriaExtensions.cs [hunk 1]
```diff
@@ -31,21 +32,33 @@ public static class ICriteriaExtensions
...
             string authViewAlias = null)
         {
-            authViewAlias = string.IsNullOrWhiteSpace(authViewAlias)
-                ? $"authView{viewName}"
-                : $"authView{authViewAlias}";
+            string entityName = $"{viewName.GetAuthorizationViewClassName()}".GetFullNameForView();
+
+            if (viewName.ContainsIgnoreCase("EducationOrganizationIdToEducationOrganizationId"))
+            {
+                authViewAlias = string.IsNullOrWhiteSpace(authViewAlias) ? $"authTable{viewName}" : $"authTable{authViewAlias}";
```
Comment (blmeyers) [inline Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/ICriteriaExtensions.cs:45]: This refactored logic seems strange.  The only thing we seem to be changing is the passed in entityName, so why is the authViewAlias calculation also sprinkled around?  Can't we just default authViewAlias (based on existence of authViewAlias parameter), default entityName based on view, and just override the nHibernate name for EducationOrganizationToEducationOrganizationId to use this new GetFullNameForTable.

File: Application/EdFi.Ods.Common/Conventions/SystemConventions.cs [hunk 2]
```diff
@@ -30,11 +38,12 @@ public static string GetAggregateExtensionBagName(this AssociationView associati
         /// </summary>
         /// <param name="associationView"></param>
         /// <returns></returns>
+
```
Comment (blmeyers) [inline Application/EdFi.Ods.Common/Conventions/SystemConventions.cs:41]: Random new line.

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Properties/launchSettings.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
   "profiles": {
     "EdFi.CodeGen - SQLServer": {
       "commandName": "Project",
-      "commandLineArgs": "-e SQLServer"
+      "commandLineArgs": "-e SQLServer -r C:\\oss-workspace\\ -i -t"
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Properties/launchSettings.json:5]: Don't hard code your workspace in here, and preferably don't default to use plugins.  See also discussion around the code gen changes and your new parameter.

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs [hunk 4]
```diff
@@ -14,16 +14,30 @@ public class Options
...
 
-        [Option("extensionPaths", Required = false, HelpText = "Array of paths for the Extension location to determine if plugin assemblies are included",Separator =',' )]
+        [Option(
+            "extensionPaths", Required = false,
+            HelpText = "Array of paths for the Extension location to determine if plugin assemblies are included",
+            Separator = ',')]
         public IEnumerable<string> ExtensionPaths { get; set; }
+
+        [Option(
+            't', "tupletableFromDatabase", Required = false, HelpText = "Request tuple table metadata from an existing empty database.",
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs:39]: I think overall this has ended up a lot more complicated than we want. Our goal is to reduce dependencies on the database where we can, particularly in code gen.  You've had to make major changes to code gen and introduced a lot of dependencies, all to generate a mapping with just two columns.  I'm fine with the approach you're taking of a separate entity and nHibernate mapping file to maintain access for this tuple table, but my recommendation would for them to just be hard coded, and not try to generate them.  This table is fairly stable, so I'm not sure we need all this complexity to generate it.  On top of that even with generation you're still requiring a very manual step to generate it, so I don't think its paying off.  I'd rip all this code gen for those artifacts and just have them checked in and we keep them in sync if we have to change the tuple table down the road.  CC: @vimayya

**Review by blmeyers — CHANGES_REQUESTED** [review]

File: Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/ICriteriaExtensions.cs [hunk 5]
```diff
@@ -31,21 +32,33 @@ public static class ICriteriaExtensions
             Junction whereJunction,
             IDictionary<string, object> parameters,
             string viewName,
-            string joinPropertyName,
+            string aggregateRootPropertyName,
+            string authViewPropertyName,
```
[the remark this answers is not in the document]
Comment (semalaiappan) [reply Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/ICriteriaExtensions.cs:37]: I updated it . Thanks

File: Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/ICriteriaExtensions.cs (hunk 1 above)
Comment (semalaiappan) [reply Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/ICriteriaExtensions.cs]: I updated it now

File: Application/EdFi.Ods.Common/Conventions/SystemConventions.cs (hunk 2 above)
Comment (semalaiappan) [reply Application/EdFi.Ods.Common/Conventions/SystemConventions.cs]: I removed it

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Properties/launchSettings.json (hunk 3 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Properties/launchSettings.json]: I removed it my changes

File: Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/RelationshipsAuthorizationFilters.cs [hunk 6]
```diff
@@ -64,7 +64,7 @@ public static class RelationshipsAuthorizationFilters
                         @" {newAlias1} 
                             WHERE {newAlias1}.SourceEducationOrganizationId IN (:SourceEducationOrganizationId))",
                         (c, w, p, jt) => c.ApplyJoinFilter(
-                            w, p, "StudentUSIToEducationOrganizationIdThroughEdOrgAssociation", "StudentUSI", "SourceEducationOrganizationId", jt),
+                            w, p, "LocalEducationAgencyIdToStudentUSIThroughEdOrgAssociation", "StudentUSI", "StudentUSI", "LocalEducationAgencyId", jt),
```
Comment (blmeyers) [inline Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/RelationshipsAuthorizationFilters.cs:67]: Why this change?  I understand needing to update this to add the new StudentUSI parameter, but why the changes to the view name and column names?  Wasn't this working correctly before?  This and the SchoolId view changes originally came in with https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/commit/313233260d911978e1f32a3364e3ef6abada64fa

File: Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/RelationshipsAuthorizationFilters.cs [hunk 7]
```diff
@@ -96,9 +96,9 @@ public static class RelationshipsAuthorizationFilters
                         WHERE {newAlias1}.SourceEducationOrganizationId IN (:SourceEducationOrganizationId))",
                         @"{currentAlias}.StudentUSI IN (
                         SELECT {newAlias1}.StudentUSI 
-                        FROM " + "auth_StudentUSIToEducationOrganizationIdThroughEdOrgAssociation".GetFullNameForView() + @" {newAlias1} 
-                        WHERE {newAlias1}.SourceEducationOrganizationId IN (:SourceEducationOrganizationId))",
-                        (c, w, p, jt) => c.ApplyJoinFilter(w, p, "StudentUSIToEducationOrganizationIdThroughEdOrgAssociation", "StudentUSI", "SchoolId", jt),
+                        FROM " + "auth_SchoolIdToStudentUSIThroughEdOrgAssociation".GetFullNameForView() + @" {newAlias1} 
+                        WHERE {newAlias1}.SchoolId IN (:SchoolId))",
+                        (c, w, p, jt) => c.ApplyJoinFilter(w, p, "SchoolIdToStudentUSIThroughEdOrgAssociation", "StudentUSI", "StudentUSI", "SchoolId", jt),
```
Comment (blmeyers) [inline Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/RelationshipsAuthorizationFilters.cs:101]: See note above about changes around ThroughEdOrgAssociation.

File: Application/EdFi.Ods.Standard/Artifacts/Metadata/DatabaseViews.generated.json [hunk 8]
```diff
@@ -954,40 +954,22 @@
     ]
   },
   {
-    "SchemaOwner": "interop",
-    "Name": "OperationalContextSupport",
+    "SchemaOwner": "auth",
+    "Name": "EducationOrganizationIdToEducationOrganizationId",
```
Comment (blmeyers) [inline Application/EdFi.Ods.Standard/Artifacts/Metadata/DatabaseViews.generated.json:958]: It looks like we lost the OperationContextSupport view as part of these updates.  Its probably fine for now since we're not using that view actively from the API, but was this intentional?

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Database/DatabaseSchema/AuthorizationDatabaseTableViewsProvider.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            }
+
+            using (var connection = _databaseConnectionProvider.CreateConnection(_connectionString))
+            {
+                connection.Open();
+
+                using (var dr = new DatabaseReader(connection))
+                {
+                    dr.CommandTimeout = 60;
+                    var dataTupleTable = dr.Table("EducationOrganizationIdToEducationOrganizationId");
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Database/DatabaseSchema/AuthorizationDatabaseTableViewsProvider.cs:48]: Preferably move this line down to just above where you're adding it to the views list.  Its generally preferable to keep variable setup as close as you can to where you're using it, rather than initializing it up here, setting up all the view data, then coming back to use the table data.

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/DatabaseViews.cs [hunk 10]
```diff
@@ -11,9 +11,9 @@ namespace EdFi.Ods.CodeGen.Generators
 {
     public class DatabaseViews : GeneratorBase
     {
-        private readonly IViewsProvider _viewsProvider;
+        private readonly IAuthorizationDatabaseTableViewsProvider _viewsProvider;
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/DatabaseViews.cs:14]: Rename of the interface is great, but update the field and parameter on the constructor to also take into account the rename.

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/DatabaseViews.cs [hunk 11]
```diff
@@ -38,7 +38,8 @@ protected override object Build()
                         }),
                 });
 
-            return new {Views = JsonConvert.SerializeObject(views, Formatting.Indented)};
+            return new {Views = JsonConvert.SerializeObject(views, Formatting.Indented ,
+                new JsonSerializerSettings {ReferenceLoopHandling = ReferenceLoopHandling.Ignore})};
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/DatabaseViews.cs:42]: What's driving this change?

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/DatabaseViews.cs [hunk 12]
```diff
@@ -38,7 +38,8 @@ protected override object Build()
                         }),
                 });
 
-            return new {Views = JsonConvert.SerializeObject(views, Formatting.Indented)};
+            return new {Views = JsonConvert.SerializeObject(views, Formatting.Indented ,
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/DatabaseViews.cs:41]:
```suggestion
            return new {Views = JsonConvert.SerializeObject(views, Formatting.Indented,
```

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntitiesForQueriesViews.cs [hunk 13]
```diff
@@ -19,10 +20,10 @@ namespace EdFi.Ods.CodeGen.Generators
     public class EntitiesForQueriesViews : GeneratorBase
     {
         private const object NotRendered = null;
-        private readonly IViewsProvider _viewsProvider;
+        private readonly IAuthorizationDatabaseTableViewsProvider _viewsProvider;
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntitiesForQueriesViews.cs:23]: Same here, rename the field and constructor param.

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntitiesForQueriesViews.cs [hunk 14]
```diff
@@ -64,7 +65,7 @@ string GetCSharpNullSuffix(DatabaseColumn c)
                         TableName = view.Name,
                         SchemaName = view.SchemaOwner,
                         Properties = view.Columns.OrderBy(c => c.Name).Select(
-                            c => new
+                            c => new
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntitiesForQueriesViews.cs:68]:
```suggestion
                            c => new
```

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappings.cs [hunk 15]
```diff
@@ -23,13 +23,13 @@ namespace EdFi.Ods.CodeGen.Generators
 {
     public class EntityOrmMappings : GeneratorBase
     {
-        private readonly IViewsProvider _viewsProvider;
+        private readonly IAuthorizationDatabaseTableViewsProvider _viewsProvider;
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappings.cs:26]: Same here, rename the field and constructor param.

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappingsForQueries.cs [hunk 16]
```diff
@@ -10,7 +10,7 @@ namespace EdFi.Ods.CodeGen.Generators
 {
     public class EntityOrmMappingsForQueries : EntityOrmMappings
     {
-        public EntityOrmMappingsForQueries(IViewsProvider viewsProvider, IDatabaseTypeTranslator databaseTypeTranslator)
+        public EntityOrmMappingsForQueries(IAuthorizationDatabaseTableViewsProvider viewsProvider, IDatabaseTypeTranslator databaseTypeTranslator)
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappingsForQueries.cs:13]: Same here, rename the field and constructor param.

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappingsForViews.cs [hunk 17]
```diff
@@ -18,10 +18,10 @@ namespace EdFi.Ods.CodeGen.Generators
     public class EntityOrmMappingsForViews : GeneratorBase
     {
         private const string NotRendered = null;
-        private readonly IViewsProvider _viewsProvider;
+        private readonly IAuthorizationDatabaseTableViewsProvider _viewsProvider;
         private readonly IDatabaseTypeTranslator _databaseTypeTranslator;
 
-        public EntityOrmMappingsForViews(IViewsProvider viewsProvider, IDatabaseTypeTranslator databaseTypeTranslator)
+        public EntityOrmMappingsForViews(IAuthorizationDatabaseTableViewsProvider viewsProvider, IDatabaseTypeTranslator databaseTypeTranslator)
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappingsForViews.cs:24]: Same here, rename the field and constructor param.

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Modules/CommandLineOverrideModule.cs [hunk 18]
```diff
@@ -35,13 +35,15 @@ protected override void Load(ContainerBuilder builder)
...
+            builder.RegisterType<AuthorizationDatabaseTableViewsProvider>()
+                .As<IAuthorizationDatabaseTableViewsProvider>()
                 .OnlyIf(x => Options.ViewsFromDatabase);
 
             builder.RegisterType<JsonViewsProvider>()
-                .As<IViewsProvider>()
+                .As<IAuthorizationDatabaseTableViewsProvider>()
                 .OnlyIf(x => !Options.ViewsFromDatabase);
+
+
```
Comment (blmeyers) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Modules/CommandLineOverrideModule.cs:46]:
```suggestion
```

**Review by blmeyers — CHANGES_REQUESTED** [review]

File: Application/EdFi.Ods.Standard/Artifacts/Metadata/DatabaseViews.generated.json (hunk 8 above)
Comment (semalaiappan) [reply Application/EdFi.Ods.Standard/Artifacts/Metadata/DatabaseViews.generated.json:958]: yes , as discussed on Friday's call , we are not using it ,so I cleaned up to reduce the maintenance overhead for not used code .

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Database/DatabaseSchema/AuthorizationDatabaseTableViewsProvider.cs (hunk 9 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Database/DatabaseSchema/AuthorizationDatabaseTableViewsProvider.cs]: I fixed it .Thanks

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/DatabaseViews.cs (hunk 10 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/DatabaseViews.cs]: I fixed it .Thanks

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntitiesForQueriesViews.cs (hunk 14 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntitiesForQueriesViews.cs]: I fixed it .Thanks

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappings.cs (hunk 15 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappings.cs]: I fixed it .Thanks

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappingsForQueries.cs (hunk 16 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappingsForQueries.cs]: I fixed it .Thanks

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappingsForViews.cs (hunk 17 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Generators/EntityOrmMappingsForViews.cs]: I fixed it .Thanks

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Modules/CommandLineOverrideModule.cs (hunk 18 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Modules/CommandLineOverrideModule.cs]: I fixed it .Thanks

File: Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/RelationshipsAuthorizationFilters.cs (hunk 6 above)
Comment (semalaiappan) [reply Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/RelationshipsAuthorizationFilters.cs]: I rebase with latest main to resolve this change

File: Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/RelationshipsAuthorizationFilters.cs (hunk 7 above)
Comment (semalaiappan) [reply Application/EdFi.Ods.Api/Security/AuthorizationStrategies/Relationships/Filters/RelationshipsAuthorizationFilters.cs]: I rebase with latest main to resolve this change

**Review by blmeyers — APPROVED** [review]
